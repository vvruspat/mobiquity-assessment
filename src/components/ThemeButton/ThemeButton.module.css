.theme_button {
    width: 24px;
    height: 24px;
    padding: 0;
    background: none;
}