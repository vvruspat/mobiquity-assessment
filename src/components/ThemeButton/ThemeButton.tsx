import { useCallback, useEffect, useState } from "react";
import { Button } from "../Button";
import { ReactComponent as LightIcon } from "./assets/light.svg";
import { ReactComponent as DarkIcon } from "./assets/dark.svg";
import s from "./ThemeButton.module.css";

// If theme not setted to localstorage we are checking system preferences
const defaultTheme =
  (localStorage?.getItem("theme") as "dark" | "light" | undefined) ??
  /** Optional chaining for tests suites */
  (window?.matchMedia?.("(prefers-color-scheme: dark)").matches
    ? "dark"
    : "light");

export const ThemeButton = () => {
  const [mode, setMode] = useState<"dark" | "light">(defaultTheme);

  const onThemeButtonClick = useCallback(() => {
    setMode((prevMode) => (prevMode === "dark" ? "light" : "dark"));
  }, []);

  useEffect(() => {
    document.body.setAttribute("data-theme", mode);
    localStorage?.setItem("theme", mode);
  }, [mode]);

  return (
    <Button onClick={onThemeButtonClick} mode="icon" className={s.theme_button}>
      {mode === "dark" ? <LightIcon /> : <DarkIcon />}
    </Button>
  );
};
