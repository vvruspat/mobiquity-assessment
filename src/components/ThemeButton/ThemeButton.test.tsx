import { render, screen, fireEvent } from "@testing-library/react";
import { ThemeButton } from "./ThemeButton";

test("renders theme button", () => {
  render(<ThemeButton />);

  const el = screen.getByTestId("button");

  fireEvent.click(el);
  expect(global.localStorage.getItem("theme")).toEqual("dark");
  fireEvent.click(el);
  expect(global.localStorage.getItem("theme")).toEqual("light");

  expect(el).toMatchSnapshot();
});
