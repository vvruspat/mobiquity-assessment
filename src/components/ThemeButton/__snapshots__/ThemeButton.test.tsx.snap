// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders theme button 1`] = `
<button
  class="button icon-button theme_button"
  data-testid="button"
>
  <svg>
    dark.svg
  </svg>
</button>
`;
